{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28c7d84b",
   "metadata": {},
   "source": [
    "# Into to NNCME with a Toggle Switch Example\n",
    "\n",
    "This is a detailed step-by-step guideline of implementing a toggle switch example with NNCME. \\\n",
    "A simple example of a gene expression system can be seen in `Start with a Simple Gene Expression Example.ipynb`.\\\n",
    "If you would like to explore more with NNCME, please refer to the `.py` files of other representative examples.\n",
    "\n",
    "\n",
    "## 1. Toggle Switch System\n",
    "The toggle switch system involves 6 molecular species:  genes $G_{x}$, $G_{y}$, which transcript two proteins $P_{x}$, $P_{y}$ inhibiting the gene expression of each other. There are two protein-DNA complexes $\\bar{G}_{x}$, $\\bar{G}_{y}$, with protein $P_{y}$ ($P_{x}$) bound to gene $G_{x}$ ($G_{y}$). The dimer of the protein $P_{x}$ ($P_{y}$) inhibits the activity of the gene $G_{y}$ ($G_{x}$). \n",
    "\n",
    "<table>\n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panela.svg\" width=\"200\"/> <br>\n",
    "</td> \n",
    "</table>\n",
    "\n",
    "The model can be written as the chemical reactions:\n",
    "\\begin{align}\n",
    "\\begin{split}\n",
    "G_{x}\\stackrel{s_{x}}{\\longrightarrow}G_{x}+P_{x},\\quad\n",
    "G_{y}\\stackrel{s_{y}}{\\longrightarrow}G_{y} & +P_{y},\\quad\n",
    "P_{x}\\stackrel{d_{x}}{\\longrightarrow}\\emptyset,\\quad\n",
    "P_{y}\\stackrel{d_{y}}{\\longrightarrow}\\emptyset, \\\\\n",
    "2P_{x}+G_{y}\\stackrel{b_{y}}{\\longrightarrow}\\bar{G}_{y},\\quad\n",
    "2P_{y}+G_{x}\\stackrel{b_{x}}{\\longrightarrow}\\bar{G}_{x},\\quad\n",
    "&\\bar{G}_{y}\\stackrel{u_{y}}{\\longrightarrow}2P_{x}+G_{y},\\quad\n",
    "\\bar{G}_{x}\\stackrel{u_{x}}{\\longrightarrow}2P_{y}+G_{x},\n",
    "\\end{split}\n",
    "\\end{align}\n",
    "where $s_{x}$, $s_{y}$ are the synthesis rates of the proteins, and $p_{x}$, $p_{y}$ are the degradation rates of the proteins. \n",
    "The transition rate $b_{y}$ ($b_{x}$) is the binding rate of two copies of protein $P_{X}$ ($P_{Y}$) to the $G_{y}$ ($G_{x}$), to form the complex $\\bar{G}_{y}$ ($\\bar{G}_{x}$). The unbinding of the complex $\\bar{G}_{y}$ ($\\bar{G}_{x}$) has the rate $u_{y}$ ($u_{x}$).  \n",
    "\n",
    "The stoichiometeric matrix can be written as with row $i$ represents species $i$ and column $k$ represents reaction $j$ (the order of species here is $G_{x}$,$G_{y}$,$P_{X}$,$P_{Y}$,$\\bar{G}_{x}$,$\\bar{G}_{y}$ï¼‰\n",
    "\n",
    "$$V=\n",
    "\\begin{bmatrix} \n",
    "    0 & 0 & 0 & 0 & 0 & -1 & 0 & 1 \\\\ \n",
    "    0 & 0 & 0 & 0 & -1 & 0 & 1 & 0 \\\\\n",
    "    1 & 0 & -1 & 0 & -2 & 0 & 2 & 0 \\\\\n",
    "    0 & 1 & 0 & -1 & 0 & -2 & 0 & 2 \\\\\n",
    "    0 & 0 & 0 & 0 & 0 & 1 & 0 & -1 \\\\\n",
    "    0 & 0 & 0 & 0 & 1 & 0 & -1 & 0\n",
    "\\end{bmatrix}.\n",
    "$$\n",
    "\n",
    "The concentrations of each species are given by the following ODEs: \n",
    "\n",
    "\\begin{gather}\n",
    "\\frac{dG_{x}}{dt}=-b_{x}P_{y}(P_{y}-1)G_{x}+u_{x}\\bar{G}_{x}\\\\\n",
    "\\frac{dG_{y}}{dt}=-b_{y}P_{x}(P_{x}-1)G_{y}+u_{y}\\bar{G}_{y}\\\\\n",
    "\\frac{dP_{x}}{dt}=s_{x}G_{x}-d_{x}P_{x}-2b_{x}P_{y}(P_{y}-1)G_{x}+2u_{y}\\bar{G}_{y}\\\\\n",
    "\\frac{dP_{y}}{dt}=s_{y}G_{y}-d_{y}P_{y}-2b_{y}P_{x}(P_{x}-1)G_{y}+2u_{x}\\bar{G}_{x}\\\\\n",
    "\\frac{d\\bar{G}_{x}}{dt}=b_{x}P_{y}(P_{y}-1)G_{x}-u_{x}\\bar{G}_{x}\\\\\n",
    "\\frac{d\\bar{G}_{y}}{dt}=b_{y}P_{x}(P_{x}-1)G_{y}-u_{y}\\bar{G}_{y}\n",
    "\\end{gather}\n",
    "\n",
    "The total count of the two forms for each gene is conserved: $G_{x}+\\bar{G}_{x}=1$. This conservation sets a constraint on the count of the two genes: $G_{x}=0, 1$, $G_{y}=0, 1$, and effectively reduces two variables by  $\\bar{G}_{x}=1-G_{x}$, $\\bar{G}_{y}=1-G_{y}$. Therefore, in the following code, the number of species is set 4."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71c3dfbd",
   "metadata": {},
   "source": [
    "## 2. Create a ToggleSwitch.py\n",
    "Input toggle switch system as a `.py` file. Include initial conditions, species number constraint, reaction rates and the stoichiometric matrix in the function 'rates' . What's more, function 'init' and 'Propensity' are essential for every system and function 'MaskToggleSwitch' is used to constrain the count of certain species in this system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5539cb63",
   "metadata": {},
   "outputs": [],
   "source": [
    "###ToggleSwitch.py\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "class ToggleSwitch:\n",
    "    \n",
    "    def rates(self): \n",
    "\n",
    "        self.L=4 # Number of spscies\n",
    "        IniDistri='delta' # Choose delta distribution to generate initial distribution\n",
    "        initialD=np.array([1,1,0,0]).reshape(1,self.L) # The parameter for the initial delta distribution\n",
    "        # here we set G_x=1, G_y=1, P_x=0, P_y=0\n",
    "        \n",
    "        MConstrain=np.array([2,2,self.M,self.M], dtype=int) # number constraint (not include the upper limit)\n",
    "        # here G_x,G_y=0,1 therefore set G_x<2, G_y<2 \n",
    "        # Add different number constraint for each species. If the upper limits for all species are all the same as self.M, then set Mconstrain=np.zeros(1,dtype=int)\n",
    "        conservation=np.ones(1,dtype=int) # To maintain the total number of all species, set conservation=initial_num.sum(), otherwise set conservation=np.ones(1,dtype=int).\n",
    "    \n",
    "        r=torch.zeros(8) # reaction rates, 8 is the reation number in this system\n",
    "        sx=sy=50\n",
    "        dx=dy=1\n",
    "        by=bx=1e-4\n",
    "        uy=ux=0.1\n",
    "        \n",
    "        r[0] = sx\n",
    "        r[1] = sy\n",
    "        r[2] = dx\n",
    "        r[3] = dy\n",
    "        r[4] = by\n",
    "        r[5] = bx\n",
    "        r[6] = uy\n",
    "        r[7] = ux\n",
    "\n",
    "        # Reaction matrix\n",
    "        ReactionMatLeft=torch.as_tensor([(1,0,0,0,0,1,0,0),(0,1,0,0,1,0,0,0),(0,0,1,0,2,0,0,0),(0,0,0,1,0,2,0,0)]).to(self.device)#SpeciesXReactions\n",
    "        #The reaction matrix on the left (dimension: species_num*reactions_num) (stoichiometric matrix = reaction_matrix_left-reaction_matrix_right) \n",
    "        ReactionMatRight=torch.as_tensor([(1,0,0,0,0,0,0,1),(0,1,0,0,0,0,1,0),(1,0,0,0,0,0,2,0),(0,1,0,0,0,0,0,2)]).to(self.device)#SpeciesXReactions\n",
    "        #The reaction matrix on the right (dimension: species_num*reactions_num) (stoichiometric matrix = reaction_matrix_left-reaction_matrix_right) \n",
    "            \n",
    "        return IniDistri,initialD,r,ReactionMatLeft,ReactionMatRight,MConstrain,conservation\n",
    "    \n",
    "    def __init__(self, *args, **kwargs):\n",
    "        super().__init__()\n",
    "        #self.n = kwargs['n']\n",
    "        self.L = kwargs['L']\n",
    "        self.M = kwargs['M']\n",
    "        self.bits = kwargs['bits']  \n",
    "        self.device = kwargs['device']\n",
    "        self.MConstrain = kwargs['MConstrain']\n",
    "        # self.Para = kwargs['Para']\n",
    "        self.IniDistri = kwargs['IniDistri']\n",
    "    \n",
    "    # It is used to constrain the count of certain species. \n",
    "    # For example, DNA typically only has the count of 0 or 1 inside a cell. \n",
    "    # The \"Mask\" function allows only the reactions with such a proper count to occur.\n",
    "    def MaskToggleSwitch(self,SampleNeighbor1D_Win,WinProd):\n",
    "        Mask1=torch.ones_like(WinProd)\n",
    "        Gx=SampleNeighbor1D_Win[:,0,:] #Gx for different reactions: the second dimension of SampleNeighbor1D_Win is the label of species\n",
    "        Gy=SampleNeighbor1D_Win[:,1,:] #Gy for different reactions: the second dimension of SampleNeighbor1D_Win is the label of species\n",
    "        Gx1=1-Gx\n",
    "        Gy1=1-Gy\n",
    "        Mask1[Gx[:,0]!=1,0]=0 #The second dimension of Gx and Mask1 is the label of reactions\n",
    "        Mask1[Gx[:,5]!=1,5]=0\n",
    "        Mask1[Gy[:,1]!=1,1]=0\n",
    "        Mask1[Gy[:,4]!=1,4]=0\n",
    "        Mask1[Gx1[:,7]!=1,7]=0\n",
    "        Mask1[Gy1[:,6]!=1,6]=0\n",
    "        \n",
    "        return Mask1\n",
    "    \n",
    "    def Propensity(self,Win,Wout,cc,SampleNeighbor1D_Win,SampleNeighbor1D,NotHappen_in_low,NotHappen_in_up,NotHappen_out_low,NotHappen_out_up):\n",
    "        WinProd=torch.prod(Win,1)\n",
    "        Mask1=self.MaskToggleSwitch(SampleNeighbor1D_Win,WinProd)\n",
    "        Propensity_in=WinProd*Mask1*cc\n",
    "        WoutProd=torch.prod(Wout,1)\n",
    "        Mask1=self.MaskToggleSwitch(SampleNeighbor1D,WoutProd)\n",
    "        Propensity_out=WoutProd*Mask1*cc\n",
    "        \n",
    "        return Propensity_in,Propensity_out"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60ef4ba7",
   "metadata": {},
   "source": [
    "## 3. Implement the code after providing the ToggleSwitch.py file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae687ec4",
   "metadata": {},
   "source": [
    "(a) Server: Use a `ToggleSwitch.sh` to input the hyperparameters from the shell. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87f6a9f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH --job-name=ToggleSwitch-rnn-1-32-M80-Tstep8001-dt0.005\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --nodes=1\n",
    "#SBATCH --partition=v100\n",
    "#SBATCH --gres=gpu:1\n",
    "#SBATCH -o out/%j.out\n",
    "#SBATCH -e out/%j.out\n",
    "#SBATCH -t 13-2:00:00\n",
    "\n",
    "# load the environment\n",
    "module purge\n",
    "\n",
    "CUDA_LAUNCH_BLOCKING=1 python3 MasterEq.py --L 4 --M 80 --Model 'ToggleSwitch' --net 'rnn' --lossType 'kl' --max_stepAll 5000 --max_stepLater 100 --lr 0.001 --net_depth 1 --net_width 32 --print_step 40 --batch_size 1000 --Tstep 8001 --delta_t 0.005 --cuda 0 --dtype float64\n",
    "\n",
    "#exit 0\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa9b2677",
   "metadata": {},
   "source": [
    "(b) PC users (Windows): Run `MasterEq.py` after necessary changes according to the toggle switch example (importing the model class, adjusting the hyperparameters and adding model command)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7dd3171",
   "metadata": {},
   "outputs": [],
   "source": [
    "###MasterEq.py\n",
    "from args import args\n",
    "import numpy as np\n",
    "from main import Test\n",
    "\n",
    "###Add models----------------------------------\n",
    "from ToggleSwitch import ToggleSwitch  #from ToggleSwitch.py file import the ToggleSwitch class\n",
    "from EarlyLife import EarlyLife\n",
    "from Epidemic import Epidemic\n",
    "from cascade1 import cascade1\n",
    "from cascade1_inverse import cascade1_inverse\n",
    "from cascade2 import cascade2\n",
    "from cascade3 import cascade3\n",
    "from BirthDeath import BirthDeath\n",
    "from GeneExp import GeneExp\n",
    "from AFL import AFL\n",
    "\n",
    "##Set parameters-------------------------------\n",
    "###Initialize parameters: otherwise the parameters are specified in init_out_dir-> args.py\n",
    "args.Model='ToggleSwitch' #Model name\n",
    "args.L=4#Species number\n",
    "args.M=int(80) #Upper limit of the molecule number: it is adjustable and can be indicated by doing a few Gillespie simulation. \n",
    "args.batch_size=1000 #Number of batch samples\n",
    "args.Tstep=8001# Time step of iterating the chemical master equation\n",
    "args.delta_t=0.0005 #Time step length of iterating the chemical master equation, depending on the reaction rates\n",
    "\n",
    "args.net ='rnn'\n",
    "args.max_stepAll=5000 #Maximum number of steps first time step (usually larger to ensure the accuracy)\n",
    "args.max_stepLater=100 #Maximum number of steps of later time steps\n",
    "args.net_depth=1 # including output layer and not input data layer\n",
    "args.net_width=32\n",
    "args.d_model=16# transformer\n",
    "args.d_ff=32# transformer\n",
    "args.n_layers=2# transformer\n",
    "args.n_heads=2# transformer\n",
    "args.lr=0.001\n",
    "args.binary=False\n",
    "args.AdaptiveT=False\n",
    "args.AdaptiveTFold=5\n",
    "args.print_step=20\n",
    "args.saving_data_time_step=[0,1e2,5e2,2e3,1e4,2e4,5e4,1e5,1.5e5,2e5,2.5e5,3e5,3.5e5,4e5,5e5,6e5,7e5,8e5,9e5,1e6] #To save data at which time steps (give in a list)\n",
    "args.training_loss_print_step=[0,1,2,101,1001,2e3,1e4,1e5,2e5,3e5,4e5,5e5] #To print training loss at which time steps (give in a list)\n",
    "\n",
    "###Default parameters:\n",
    "args.bias=True\n",
    "args.bits=1\n",
    "if args.binary:\n",
    "    args.bits=int(np.ceil(np.log2(args.M)))\n",
    "args.Percent=0.2         \n",
    "args.clip_grad=1\n",
    "args.dtype='float64'\n",
    "args.epsilon=1e-30#initial probability of zero species number\n",
    "args.lr_schedule=False#True\n",
    "\n",
    "###Add model command-------------------------------\n",
    "if args.Model=='ToggleSwitch':\n",
    "    model = ToggleSwitch(**vars(args))   \n",
    "if args.Model=='EarlyLife':\n",
    "    model = EarlyLife(**vars(args))   \n",
    "if args.Model=='Epidemic':\n",
    "    model = Epidemic(**vars(args))   \n",
    "if args.Model=='cascade1': \n",
    "    model = cascade1(**vars(args)) \n",
    "if args.Model=='cascade1_inverse':\n",
    "    model = cascade1_inverse(**vars(args)) \n",
    "if args.Model=='cascade2':\n",
    "    model = cascade2(**vars(args))    \n",
    "if args.Model=='cascade3':\n",
    "    model = cascade3(**vars(args))    \n",
    "if args.Model=='BirthDeath':\n",
    "    model = BirthDeath(**vars(args))   \n",
    "if args.Model=='GeneExp':\n",
    "    model = GeneExp(**vars(args))   \n",
    "if args.Model=='AFL':\n",
    "    model = AFL(**vars(args)) \n",
    "    \n",
    "#Run model-----------------------------------------        \n",
    "if __name__ == '__main__':\n",
    "    Test(model)    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "909a2c39",
   "metadata": {},
   "source": [
    "## 4. Optional: Run Gillespie simulation\n",
    "Gillespie algorithm, also known as the stochastic simulation algorithm or the kinetic Monte Carlo, simulates trajectories to generate statistics of relevant variables. To evaluate the accuracy of the learnt distribution by the VAN, we can compare the resultant marginal distribution of one species with those from Gillespie algorithm."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6578122e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import biocircuits\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def ToggleSwitch_propensity(\n",
    "    propensities, population, t, sx, sy, dx, dy, by, bx, uy, ux\n",
    "):\n",
    "    #species\n",
    "    Gx, Gy, Px, Py = population  \n",
    "    Gx1=1-Gx\n",
    "    Gy1=1-Gy\n",
    "    #propensities\n",
    "    propensities[0] = sx*Gx\n",
    "    propensities[1] = sy*Gy\n",
    "    propensities[2] = dx*Px\n",
    "    propensities[3] = dy*Py\n",
    "    propensities[4] = by*Px*Px*Gy\n",
    "    propensities[5] = bx*Py*Py*Gx\n",
    "    propensities[6] = uy*Gy1\n",
    "    propensities[7] = ux*Gx1\n",
    "\n",
    "#the stoichiometric matrix   \n",
    "ToggleSwitch_update = np.array(\n",
    "    [\n",
    "        [ 0, 0, 1, 0], \n",
    "        [ 0, 0, 0, 1],\n",
    "        [ 0, 0,-1, 0],\n",
    "        [ 0, 0, 0,-1], \n",
    "        [ 0,-1,-2, 0], \n",
    "        [-1, 0, 0,-2],\n",
    "        [ 0, 1, 2, 0],\n",
    "        [ 1, 0, 0, 2]\n",
    "    ],dtype=int)\n",
    "\n",
    "#the reaction rates \n",
    "sx=sy=50\n",
    "dx=dy=1\n",
    "by=bx=1e-4\n",
    "uy=ux=0.1\n",
    "\n",
    "ToggleSwitch_args = (sx, sy, dx, dy, by, bx, uy, ux)\n",
    "\n",
    "#initial number of species \n",
    "Gx0=0\n",
    "Gy0=0\n",
    "Px0=0\n",
    "Py0=0\n",
    "ToggleSwitch_pop_0 = np.array([Gx0,Gy0,Px0,Py0], dtype=float) # follow VAN's learnt initial number\n",
    "\n",
    "#simulation time length\n",
    "T=40\n",
    "time_points = np.linspace(0, T, T+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e05c0871",
   "metadata": {},
   "outputs": [],
   "source": [
    "Run=1 #run Gillespie or not\n",
    "times=1000 #Gillespie simulation times\n",
    "\n",
    "out_filename = 'ToggleSwitch_times'+str(times)+'_T'+str(T)+'_dis'+str(Gx0)+'_'+str(Gy0)+str(Px0)+'_'+str(Py0) #filename to save\n",
    "if Run==1:\n",
    "\n",
    "    Gx_total=np.empty(shape=(0,len(time_points)))#to save the time evolution of Gx number in each simulation (dimension: times*time_points)\n",
    "    Gx1_total=np.empty(shape=(0,len(time_points)))#to save the time evolution of \\bar{Gx} number in each simulation (dimension: times*time_points)\n",
    "    Gy_total=np.empty(shape=(0,len(time_points)))#to save the time evolution of Gy number in each simulation (dimension: times*time_points)\n",
    "    Gy1_total=np.empty(shape=(0,len(time_points)))#to save the time evolution of \\bar{Gy} number in each simulation (dimension: times*time_points)\n",
    "    Px_total=np.empty(shape=(0,len(time_points)))#to save the time evolution of Px number in each simulation (dimension: times*time_points)\n",
    "    Py_total=np.empty(shape=(0,len(time_points)))#to save the time evolution of Py number in each simulation (dimension: times*time_points)\n",
    "\n",
    "    for i in range(times):\n",
    "        \n",
    "        # Perform the Gillespie simulation\n",
    "        pop = biocircuits.gillespie_ssa(\n",
    "            ToggleSwitch_propensity,\n",
    "            ToggleSwitch_update,\n",
    "            ToggleSwitch_pop_0,\n",
    "            time_points,\n",
    "            args=ToggleSwitch_args,\n",
    "        )\n",
    "        \n",
    "        Gx_total=np.row_stack((Gx_total,pop[0,:,0]))\n",
    "        Gx1_total=np.row_stack((Gx1_total,1-pop[0,:,0]))\n",
    "        Gy_total=np.row_stack((Gy_total,pop[0,:,1]))\n",
    "        Gy1_total=np.row_stack((Gy1_total,1-pop[0,:,1]))\n",
    "        Px_total=np.row_stack((Px_total,pop[0,:,2]))\n",
    "        Py_total=np.row_stack((Py_total,pop[0,:,3]))\n",
    "\n",
    "        if i<3: #Plot fist three simulations\n",
    "            plt.rc('font', size=16)\n",
    "            plt.figure()\n",
    "            plt.plot(time_points ,pop[0,:,0])\n",
    "            plt.plot(time_points ,pop[0,:,1])\n",
    "            plt.plot(time_points ,pop[0,:,2])\n",
    "            plt.plot(time_points ,pop[0,:,3])            \n",
    "            plt.xlabel(\"time\")\n",
    "            plt.ylabel(\"species number\")\n",
    "            plt.grid()\n",
    "            plt.show()\n",
    "        \n",
    "    np.savez('{}'.format(out_filename),np.array(times),np.array(time_points),Gx_total,Gx1_total,Gy_total,Gy1_total,Px_total,Py_total) #save data\n",
    "else: #load existing data file\n",
    "    data=np.load(out_filename+'.npz')\n",
    "    print(list(data))    \n",
    "    time_points = data['arr_1']\n",
    "    Gx_total = data['arr_2']\n",
    "    Gx1_total = data['arr_3']\n",
    "    Gy_total = data['arr_4']\n",
    "    Gy1_total = data['arr_5']\n",
    "    Px_total = data['arr_6']\n",
    "    Py_total = data['arr_7']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f21aad72",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot Gillespie Result\n",
    "plt.rc('font', size=16)\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k')\n",
    "fig, axes = plt.subplots(2,1)\n",
    "fig.tight_layout()\n",
    "ax = plt.subplot(1,1, 1)\n",
    "from matplotlib import cm\n",
    "jet = cm.get_cmap('jet')\n",
    "Number=9\n",
    "jet_12_colors = jet(np.linspace(0, 1, Number))\n",
    "for Species in np.arange(4):\n",
    "    if Species==0:\n",
    "        plt.errorbar(time_points, np.mean(Gx_total,0),  yerr=np.std(Gx_total,0), color=jet_12_colors[Species*2+1,:])    \n",
    "    if Species==1:\n",
    "        plt.errorbar(time_points, np.mean(Gy_total,0),  yerr=np.std(Gy_total,0), color=jet_12_colors[Species*2+1,:])    \n",
    "    if Species==2:\n",
    "        plt.errorbar(time_points, np.mean(Px_total,0),  yerr=np.std(Px_total,0), color=jet_12_colors[Species*2+1,:])  \n",
    "    if Species==3:\n",
    "        plt.errorbar(time_points, np.mean(Py_total,0),  yerr=np.std(Py_total,0), color=jet_12_colors[Species*2+1,:])  \n",
    "\n",
    "plt.xlabel('Time')\n",
    "plt.ylabel('Species Number')\n",
    "plt.legend()\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.figure()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63e43b35",
   "metadata": {},
   "source": [
    "## 5. Plot Results\n",
    "Plot the result of the genetic toggle switch. Relevant details can be referred in the manuscript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acc366a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data of the VAN and Gillespie\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import cm\n",
    "import math\n",
    "import pandas as pd\n",
    "from matplotlib.colors import ListedColormap, LinearSegmentedColormap\n",
    "import matplotlib as mpl\n",
    "\n",
    "# choose colors for different species\n",
    "from matplotlib import cm\n",
    "jet = cm.get_cmap('jet')\n",
    "Number=15\n",
    "jet_12_colors = jet(np.linspace(0, 1, Number))\n",
    "color1=jet_12_colors[0,:]\n",
    "color2=jet_12_colors[4,:]\n",
    "color3=jet_12_colors[12,:]\n",
    "color4=jet_12_colors[8,:]\n",
    "\n",
    "out_filename = 'ToggleSwitch/ToggleSwitch_times1000_T40_merge' # Gillespie data path\n",
    "data=np.load(out_filename+'.npz')\n",
    " \n",
    "times = data['arr_0']   \n",
    "time_points = data['arr_1']\n",
    "Gx_total = data['arr_2']\n",
    "Gx1_total = data['arr_3']\n",
    "Gy_total = data['arr_4']\n",
    "Gy1_total = data['arr_5']\n",
    "Px_total = data['arr_6']\n",
    "Py_total = data['arr_7']\n",
    "\n",
    "#############################\n",
    "AdaptiveT=0\n",
    "Name='ToggleSwitch/Data-ToggleSwitch_1_32_M80_T8002_dt0.005_batch1000_epoch100' # VAN data path\n",
    "data=np.load(str(Name)+'.npz', allow_pickle=True) \n",
    "\n",
    "argsSave = data['arr_1']\n",
    "Tstep=argsSave[0]#1001\n",
    "delta_t=argsSave[1]#0.05\n",
    "L=argsSave[2]\n",
    "print_step= argsSave[7]\n",
    "\n",
    "SampleSum=data['arr_5']\n",
    "delta_T= data['arr_6']\n",
    "if AdaptiveT: TimePoins=np.cumsum(delta_T)[np.arange(SampleSum.shape[0])*print_step]\n",
    "else: TimePoins=np.cumsum(delta_T)[np.arange(SampleSum.shape[0])*print_step]*delta_t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f13822c0",
   "metadata": {},
   "source": [
    "### 5.1 The time evolution of the average counts of species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63051b69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The time evolution of the average counts for the genes and proteins,  from the VAN (dots) and the Gillespie simulation (lines). \n",
    "#####curve-----------------------------------\n",
    "plt.rc('font', size=54)\n",
    "markersize0=12\n",
    "step=8\n",
    "\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=8)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "plt.plot(time_points,np.mean(Gx_total,0),linewidth=6,color=color1,label='$G_x$')\n",
    "plt.plot(TimePoins[::step],np.mean(SampleSum[:,:,0][::step],axis=1),\n",
    "          marker='o',linestyle = 'None',color=color1,markersize=markersize0)\n",
    "\n",
    "plt.plot(time_points,np.mean(Gy_total,0),linewidth=6,color=color2,label='$G_y$')\n",
    "plt.plot(TimePoins[::step],np.mean(SampleSum[:,:,1][::step],axis=1),\n",
    "          marker='o',linestyle = 'None',color=color2,markersize=markersize0)\n",
    "plt.xlabel(\"Time (hr)\")\n",
    "# plt.title(\"Average Number\")\n",
    "plt.ylim([0,1.05])\n",
    "plt.yticks([0,0.5,1])\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.legend(numpoints=2,handletextpad=0.2,loc='lower right')\n",
    "plt.title('Average Count',fontsize=64)\n",
    "# plt.show()\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panela1.svg', bbox_inches=\"tight\", dpi=400)\n",
    "\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=8)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "plt.plot(time_points,np.mean(Py_total,0),linewidth=6,color=color4,label='$P_y$')\n",
    "plt.plot(TimePoins[::step],np.mean(SampleSum[:,:,3][::step],axis=1),\n",
    "          marker='o',linestyle = 'None',color=color4,markersize=markersize0)\n",
    "plt.plot(time_points,np.mean(Px_total,0),linewidth=6,color=color3,label='$P_x$')\n",
    "plt.plot(TimePoins[::step],np.mean(SampleSum[:,:,2][::step],axis=1),\n",
    "          marker='o',linestyle = 'None',color=color3,markersize=markersize0)\n",
    "plt.xlabel(\"Time (hr)\")\n",
    "plt.title(\"Average Count\")\n",
    "plt.ylim(top=42)\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.legend(numpoints=2,handletextpad=0.2)\n",
    "# plt.show()\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panela2.svg', bbox_inches=\"tight\", dpi=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63d8b28d",
   "metadata": {},
   "source": [
    "<table>\n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panela1.svg\" width=\"200\"/> <br>\n",
    "</td> \n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panela2.svg\" width=\"200\"/> <br>\n",
    "</td> \n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e092e22a",
   "metadata": {},
   "source": [
    "### 5.2 Comparison of the means and standard deviations of the counts of species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca0530ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comparison of the means and standard deviations of the counts of the genes and proteins between the VAN and the Gillespie simulation, at time points t = 0, 1, ..., 39, 40.\n",
    "####Gene-------------\n",
    "plt.rc('font', size=56)\n",
    "legendsize=50\n",
    "T=40\n",
    "NumTimePoints=1\n",
    "markersize0=20\n",
    "transparency=1-np.linspace(0, 0.99, int(T/NumTimePoints))\n",
    "Range=[0,1]\n",
    "\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=8)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "# Range=1#math.ceil(max(np.mean(Gx_total,0)))\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.mean(Gx_total[:,int(tt)],0),np.mean(SampleSum[:,:,0],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color1,markersize=markersize0,label='$G_x$')\n",
    "    else:\n",
    "        plt.plot(np.mean(Gx_total[:,int(tt)],0),np.mean(SampleSum[:,:,0],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color1,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(Range,Range,linewidth=4,color='black')\n",
    "\n",
    "# Range=math.ceil(max(np.mean(Gy_total,0)))\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.mean(Gy_total[:,int(tt)],0),np.mean(SampleSum[:,:,1],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color2,markersize=markersize0,label='$G_y$')\n",
    "    else:\n",
    "        plt.plot(np.mean(Gy_total[:,int(tt)],0),np.mean(SampleSum[:,:,1],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color2,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(Range,Range,linewidth=4,color='black')\n",
    "\n",
    "\n",
    "plt.xlabel('Gillespie')\n",
    "plt.ylabel('VAN')\n",
    "plt.title('Mean')\n",
    "# plt.xticks([0.4,0.6,0.8])\n",
    "# plt.yticks([0,30])\n",
    "plt.legend(fontsize=legendsize,loc='best',handletextpad=0.2)\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panelb11.svg', bbox_inches=\"tight\", dpi=400)\n",
    "\n",
    "\n",
    "#####std----------------------------------\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=8)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "Range=[0,1]\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.std(Gx_total[:,int(tt)],0),np.std(SampleSum[:,:,0],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color1,markersize=markersize0,label='$G_x$')\n",
    "    else:\n",
    "        plt.plot(np.std(Gx_total[:,int(tt)],0),np.std(SampleSum[:,:,0],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color1,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(Range,Range,linewidth=4,color='black')\n",
    "\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.std(Gy_total[:,int(tt)],0),np.std(SampleSum[:,:,1],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color2,markersize=markersize0,label='$G_y$')\n",
    "    else:\n",
    "        plt.plot(np.std(Gy_total[:,int(tt)],0),np.std(SampleSum[:,:,1],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color2,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(Range,Range,linewidth=4,color='black')\n",
    "\n",
    "plt.xlabel('Gillespie')\n",
    "plt.ylabel('VAN')\n",
    "plt.title('Std')\n",
    "\n",
    "plt.legend(fontsize=legendsize,loc='best',handletextpad=0.2)\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panelb12.svg', bbox_inches=\"tight\", dpi=400)\n",
    "\n",
    "####Protein-------------\n",
    "plt.rc('font', size=56)\n",
    "legendsize=50\n",
    "T=40\n",
    "NumTimePoints=1\n",
    "markersize0=20\n",
    "transparency=1-np.linspace(0, 0.99, int(T/NumTimePoints))\n",
    "\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=8)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "Range=50#math.ceil(max(np.mean(Py_total,0)))\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.mean(Py_total[:,int(tt)],0),np.mean(SampleSum[:,:,3],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color4,markersize=markersize0,label='$P_y$')\n",
    "    else:\n",
    "        plt.plot(np.mean(Py_total[:,int(tt)],0),np.mean(SampleSum[:,:,3],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color4,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(np.array([0,Range]),np.array([0,Range]),linewidth=4,color='black')\n",
    "\n",
    "Range=math.ceil(max(np.mean(Px_total,0)))\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.mean(Px_total[:,int(tt)],0),np.mean(SampleSum[:,:,2],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color3,markersize=markersize0,label='$P_x$')\n",
    "    else:\n",
    "        plt.plot(np.mean(Px_total[:,int(tt)],0),np.mean(SampleSum[:,:,2],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color3,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(np.array([0,Range]),np.array([0,Range]),linewidth=4,color='black')\n",
    "\n",
    "\n",
    "plt.xlabel('Gillespie')\n",
    "plt.ylabel('VAN')\n",
    "plt.title('Mean')\n",
    "plt.legend(fontsize=legendsize,loc='best',handletextpad=0.2)\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panelb21.svg', bbox_inches=\"tight\", dpi=400)\n",
    "\n",
    "#####std----------------------------------\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=8)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "Range=math.ceil(max(np.std(Py_total,0)))\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.std(Py_total[:,int(tt)],0),np.std(SampleSum[:,:,3],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color4,markersize=markersize0,label='$P_y$')\n",
    "    else:\n",
    "        plt.plot(np.std(Py_total[:,int(tt)],0),np.std(SampleSum[:,:,3],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color4,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(np.array([0,Range]),np.array([0,Range]),linewidth=4,color='black')\n",
    "\n",
    "Range=25#.ceil(max(np.std(Px_total,0)))\n",
    "k=0\n",
    "for tt in np.arange(0,T,NumTimePoints):\n",
    "    if tt==0:\n",
    "        plt.plot(np.std(Px_total[:,int(tt)],0),np.std(SampleSum[:,:,2],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color3,markersize=markersize0,label='$P_x$')\n",
    "    else:\n",
    "        plt.plot(np.std(Px_total[:,int(tt)],0),np.std(SampleSum[:,:,2],axis=1)[round(tt/np.max(TimePoins)*TimePoins.shape[0])],\n",
    "                  alpha=transparency[k], marker='o',linestyle = 'None',color=color3,markersize=markersize0)\n",
    "    k=k+1    \n",
    "plt.plot(np.array([0,Range]),np.array([0,Range]),linewidth=4,color='black')\n",
    "\n",
    "plt.xlabel('Gillespie')\n",
    "plt.ylabel('VAN')\n",
    "plt.title('Std')\n",
    "plt.legend(fontsize=legendsize,loc='best',handletextpad=0.2)\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panelb22.svg', bbox_inches=\"tight\", dpi=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8799d948",
   "metadata": {},
   "source": [
    "<table>\n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panelb11.svg\" width=\"200\"/> <br>\n",
    "</td> \n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panelb12.svg\" width=\"200\"/> <br>\n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panelb21.svg\" width=\"200\"/> <br>\n",
    "</td> \n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panelb22.svg\" width=\"200\"/> <br>\n",
    "</td> \n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11bcea63",
   "metadata": {},
   "source": [
    "### 5.3 The marginal distributions of species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f9614f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The marginal distributions of the gene Gx and protein Px at time point t=40 from the Gillespie simulation and the VAN. The inset shows the Hellinger distance between the two distributions.\n",
    "#####Distribution---------------------------------------\n",
    "plt.rc('font', size=38)\n",
    "legendsize=30\n",
    "# plt.rc('lines',linewidth=4)\n",
    "T=40\n",
    "Range1=3\n",
    "Range2=8\n",
    "Range3=8\n",
    "axlinewidth=1.2\n",
    "\n",
    "Gx1=Gx_total[:,int(T)]\n",
    "Gx2=SampleSum[:,:,0][round(T/np.max(TimePoins)*TimePoins.shape[0])-1]\n",
    "Gy1=Gy_total[:,int(T)]\n",
    "Gy2=SampleSum[:,:,1][round(T/np.max(TimePoins)*TimePoins.shape[0])-1]\n",
    "Px1=Px_total[:,int(T)]\n",
    "Px2=SampleSum[:,:,2][round(T/np.max(TimePoins)*TimePoins.shape[0])-1]\n",
    "Py1=Py_total[:,int(T)]\n",
    "Py2=SampleSum[:,:,2][round(T/np.max(TimePoins)*TimePoins.shape[0])-1]\n",
    "df_Gx1=pd.DataFrame(Gx1)\n",
    "df_Gx2=pd.DataFrame(Gx2)\n",
    "pro_Gx1=df_Gx1.value_counts(normalize=True).sort_index()\n",
    "pro_Gx2=df_Gx2.value_counts(normalize=True).sort_index() \n",
    "df_Px1=pd.DataFrame(Px1)\n",
    "df_Px2=pd.DataFrame(Px2)\n",
    "pro_Px1=df_Px1.value_counts(normalize=True).sort_index()\n",
    "pro_Px2=df_Px2.value_counts(normalize=True).sort_index()\n",
    "\n",
    "HD_Gx=round(np.sqrt(1-np.sum(np.sqrt(pro_Gx1*pro_Gx2))),3)\n",
    "HD_Px=round(np.sqrt(1-np.sum(np.sqrt(pro_Px1*pro_Px2))),3)\n",
    "\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=4)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "fig.tight_layout()\n",
    "ax = plt.subplot(1,1, 1)\n",
    "weights1 = np.ones_like(Gx1)/float(len(Gx1))\n",
    "weights2 = np.ones_like(Gx2)/float(len(Gx2))\n",
    "plt.hist([Gx1,Gx2],bins=Range1,weights=[weights1,weights2],color=['darkgrey',color1],alpha=0.7,orientation='horizontal')\n",
    "plt.legend(['Gillespie','VAN'],title='$D_{HD}=$'+str(HD_Gx),title_fontsize=legendsize,fontsize=legendsize)\n",
    "plt.ylabel(\"$G_x$\")\n",
    "plt.xlabel('Probability')\n",
    "# plt.xticks([0,0.5])\n",
    "plt.title(\"$t=$\"+str(T))\n",
    "plt.ylim(top=3)\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panelc1_T'+str(T)+'.svg', bbox_inches=\"tight\", dpi=400)\n",
    "\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=4)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "fig.tight_layout()\n",
    "ax = plt.subplot(1,1, 1)\n",
    "weights1 = np.ones_like(Px1)/float(len(Px1))\n",
    "weights2 = np.ones_like(Px2)/float(len(Px2))\n",
    "plt.hist([Px1,Px2],bins=Range2,weights=[weights1,weights2],color=['darkgrey',color3],alpha=0.7,orientation='horizontal')\n",
    "plt.legend(['Gillespie','VAN'],title='$D_{HD}=$'+str(HD_Px),title_fontsize=legendsize,fontsize=legendsize)\n",
    "plt.ylabel(\"$P_x$\")\n",
    "plt.xlabel('Probability')\n",
    "plt.title(\"$t=$\"+str(T))\n",
    "plt.ylim(top=123)\n",
    "fig.set_size_inches(9, 8)\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_panelc2_T'+str(T)+'.svg', bbox_inches=\"tight\", dpi=400)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b79e72bc",
   "metadata": {},
   "source": [
    "<table>\n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panelc1_T40.svg\" width=\"200\"/> <br>\n",
    "</td> \n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_panelc2_T40.svg\" width=\"220\"/> <br>\n",
    "</td> \n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3099932c",
   "metadata": {},
   "source": [
    "### 5.4 The joint distribution of species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b4e3e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The joint distribution of the two proteins from the VAN at time point t=40, with the color bar for the probability values in the logarithmic scale.\n",
    "###hist2d------------------------------------------------\n",
    "cmpname='viridis'\n",
    "\n",
    "plt.rc('font', size=38)\n",
    "\n",
    "T0=40\n",
    "T=round(T0/np.max(TimePoins)*TimePoins.shape[0])-1\n",
    "bins=[35,35]#[int(max(SampleSum[T,:,2])),int(max(SampleSum[T,:,3]))]\n",
    "plt.figure(num=None,  dpi=400, edgecolor='k', linewidth=4)\n",
    "fig, axes = plt.subplots(1,1)\n",
    "fig.tight_layout()\n",
    "ax = plt.subplot(1,1, 1,facecolor=[68/255,1/255,80/255])\n",
    "h=plt.hist2d(SampleSum[T,:,2],SampleSum[T,:,3],bins=bins,cmap=cmpname,density=True,norm=mpl.colors.LogNorm(vmax=0.005,vmin=1e-5))\n",
    "# plt.colorbar(label='Probability')\n",
    "plt.xlabel(\"$P_x$\")\n",
    "plt.ylabel(\"$P_y$\")\n",
    "plt.title(\"$t=$\"+str(T0))\n",
    "plt.xlim(right=70)\n",
    "plt.ylim(top=70)\n",
    "fig.set_size_inches(9,8)\n",
    "plt.savefig('ToggleSwitch\\ToggleSwitch_paneld_T'+str(T0)+'.svg', bbox_inches=\"tight\", dpi=400)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1f56cd1",
   "metadata": {},
   "source": [
    "<table>\n",
    "<td> \n",
    "<img src=\"https://github.com/jiadeyu0602/CheatSheet/raw/master/ToggleSwitch_paneld_T40.svg\" width=\"300\"/> <br>\n",
    "</td> \n",
    "</table>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": false,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "581.333px",
    "left": "27px",
    "top": "111.125px",
    "width": "288px"
   },
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
